codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java

```diff
@@ -61,27 +68,92 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
      */
     @Override
     public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
-        if (msg instanceof FullHttpMessage) {
-            FullHttpMessage httpMsg = (FullHttpMessage) msg;
-            boolean hasData = httpMsg.content().isReadable();
-            boolean httpMsgNeedRelease = true;
+
+        if (msg instanceof HttpMessage) {
+            final HttpMessage httpMsg = (HttpMessage) msg;
+            boolean release = true;
             SimpleChannelPromiseAggregator promiseAggregator = null;
             try {
                 // Provide the user the opportunity to specify the streamId
-                int streamId = getStreamId(httpMsg.headers());
+                currentStreamId = getStreamId(httpMsg.headers());
 
                 // Convert and write the headers.
                 Http2Headers http2Headers = HttpUtil.toHttp2Headers(httpMsg);
                 Http2ConnectionEncoder encoder = encoder();
 
-                if (hasData) {
-                    promiseAggregator = new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
-                    encoder.writeHeaders(ctx, streamId, http2Headers, 0, false, promiseAggregator.newPromise());
-                    httpMsgNeedRelease = false;
-                    encoder.writeData(ctx, streamId, httpMsg.content(), 0, true, promiseAggregator.newPromise());
-                    promiseAggregator.doneAllocatingPromises();
+                if (httpMsg instanceof FullHttpMessage) {
+                    FullHttpMessage fullHttpMsg = (FullHttpMessage) httpMsg;
+                    boolean hasData = fullHttpMsg.content().isReadable();
+                    Http2Headers trailingHeaders = fullHttpMsg.trailingHeaders().isEmpty() ? null :
+                            HttpUtil.toHttp2Headers((LastHttpContent) fullHttpMsg);
+                    if (hasData) {
+                        promiseAggregator = new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
+                        encoder.writeHeaders(ctx, currentStreamId, http2Headers, 0, false,
+                                promiseAggregator.newPromise());
+
+                        final boolean hasTrailingHeaders = trailingHeaders != null && !trailingHeaders.isEmpty();
+                        encoder.writeData(ctx, currentStreamId, fullHttpMsg.content(), 0, !hasTrailingHeaders,
+                                promiseAggregator.newPromise());
+                        release = false;
+
+                        if (hasTrailingHeaders) {
+                            encoder.writeHeaders(ctx, currentStreamId, trailingHeaders, 0, true,
+                                    promiseAggregator.newPromise());
+                        }
+
+                        promiseAggregator.doneAllocatingPromises();
+                    } else {
+                        encoder.writeHeaders(ctx, currentStreamId, http2Headers, 0, true, promise);
```

Comment:
I think we should be able to simplify the logic a bit.  Could we do something like this:

```java
HttpContent content = null;
// Just always use the promise aggregator.
SimpleChannelPromiseAggregator promiseAggregator = new SimpleChannelPromiseAggregator(promise, ctx.channel, ctx.executor);

if (msg instanceof HttpMessage) {
    HttpMessage httpMsg = (HttpMessage) msg;

    // Write headers using the promise aggregator.
}

if (msg instanceof HttpContent) {
    HttpContent httpContent = (HttpContent) msg;

    // Write the data.

    if (msg instanceof LastHttpContent) {
        LastHttpContent lastHttpContent = (LastHttpContent) msg;

        // Write Data

        // Write trailers if provided.
    }
}

promiseAggregator.doneAllocatingPromises();
```

WDYT?